qa/workunits/rbd/compare_mirror_images.sh

```diff
@@ -0,0 +1,148 @@
+#!/bin/bash
+
+set -ex
+
+CEPH_ARGS=''
+IMG_PREFIX=image-primary
+MIRROR_IMAGE_MODE=snapshot
+MIRROR_POOL_MODE=image
+MNTPT_PREFIX=test-primary
+RBD_IMAGE_FEATURES='layering,exclusive-lock,object-map,fast-diff'
+RBD_MIRROR_INSTANCES=1
+RBD_MIRROR_MODE=snapshot
+RBD_MIRROR_USE_EXISTING_CLUSTER=1
+WORKLOAD_TIMEOUT=5m
+
+. $(dirname $0)/rbd_mirror_helpers.sh
+
+take_mirror_snapshots() {
+  local cluster=$1
+  local pool=$2
+  local image=$3
+
+  for i in {1..30}; do
+    mirror_image_snapshot $cluster $pool $image
+    sleep 3s;
+  done
+}
+
+slow_untar_workload() {
+  local tarball_src_path=$1
+  local mountpt=$2
+  local timeout=$3
+
+  cp $tarball_src_path $mountpt/kernel.tar.gz
+  # run workload that updates the data and metadata of multiple files on disk.
+  # rate limit the workload such that the mirror snapshots can be taken as the
+  # contents of the image are progressively changed by the workload.
+  timeout $timeout bash -c "zcat $mountpt/kernel.tar.gz \
+    | pv -L 256K | tar xf - -C $mountpt" || true
+}
+
+wait_for_image_removal () {
+  local cluster=$1
+  local pool=$2
+  local image=$3
+
+  for s in 1 2 4 8 8 8 8 8 8 8 8 16 16; do
+    if [[ -z $(rbd --cluster $cluster ls $pool | grep -w $image) ]]; then
+      echo "image:${image} removed from cluster:${cluster} pool:${pool}"
+      return 0
+    fi
+
+    echo "waiting for image ${image} to be removed from \
+      cluster:${cluster} pool:${pool} ..."
+    sleep $s
+  done
+
+  echo "image:${image} not removed from cluster:${cluster} pool:${pool}"
+  return 1
+}
+
+compare_demoted_promoted_mirror_snaps() {
+  local dev=$1
+  local img=${IMG_PREFIX}$2
+  local mntpt=${MNTPT_PREFIX}$2
+  local demote_md5 promote_md5
+
+  sudo umount ${mntpt}
+
+  # calculate hash before demotion of primary image
+  demote_md5=$(sudo md5sum ${dev} | awk '{print $1}')
+  sudo rbd --cluster ${CLUSTER1} device unmap -t ${RBD_DEVICE_TYPE} \
+      ${POOL}/${img}
+
+  demote_image ${CLUSTER1} ${POOL} ${img}
+  wait_for_status_in_pool_dir ${CLUSTER1} ${POOL} ${img} 'up+unknown'
```

Comment:
Fails here for `compare_mirror_images.sh` workloads
See, http://qa-proxy.ceph.com/teuthology/rraja-2024-02-05_20:53:19-rbd:mirror-main-distro-default-smithi/7547714/teuthology.log
<pre>
2024-02-05T21:45:07.311 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ rbd --cluster cluster1 mirror pool status mirror --verbose --format xml
2024-02-05T21:45:07.312 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:-:1.1: Document is empty
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:^
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:-:1.1: Document is empty
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:^
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ state=
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ state=
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ echo ''
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ grep up+unknown
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ echo ''
2024-02-05T21:45:07.313 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ grep up+unknown
2024-02-05T21:45:07.314 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ test '' '>' '2024-02-05 21:45:06'
2024-02-05T21:45:07.314 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ echo ''\''mirror pool status'\'' test failed'
2024-02-05T21:45:07.314 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:'mirror pool status' test failed
2024-02-05T21:45:07.314 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ exit 1
2024-02-05T21:45:07.314 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ test '' '>' '2024-02-05 21:45:06'
2024-02-05T21:45:07.315 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ echo ''\''mirror pool status'\'' test failed'
2024-02-05T21:45:07.315 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:'mirror pool status' test failed
2024-02-05T21:45:07.315 INFO:tasks.workunit.cluster1.client.mirror.smithi192.stderr:+ exit 1
</pre>


However, `compare_mirror_image_alternate_primary` workloads pass. See, https://pulpito.ceph.com/rraja-2024-02-05_20:53:19-rbd:mirror-main-distro-default-smithi/

Need to debug the failure

Reply:
After making the following change to `test_status_in_pool_dir()` the tests passed,
```
diff --git a/qa/workunits/rbd/rbd_mirror_helpers.sh b/qa/workunits/rbd/rbd_mirror_helpers.sh
index f7a3e5de0d9..21c4686a411 100755
--- a/qa/workunits/rbd/rbd_mirror_helpers.sh
+++ b/qa/workunits/rbd/rbd_mirror_helpers.sh
@@ -841,11 +841,8 @@ test_status_in_pool_dir()
 
     local last_update="$(sed -nEe 's/^  last_update: *(.*) *$/\1/p' ${status_log})"
     test_mirror_pool_status_verbose \
-        ${cluster} ${pool} ${image} "${state_pattern}" "${last_update}" &&
-    return 0
+        ${cluster} ${pool} ${image} "${state_pattern}" "${last_update}" || return 1
 
-    echo "'mirror pool status' test failed" >&2
-    exit 1
 }

```
https://github.com/ajarr/ceph/commit/2523ebdb81f1c79c150aa6f8e85685273cc1060e?diff=split&w=0

Tests passed here,
https://pulpito.ceph.com/rraja-2024-02-07_03:00:41-rbd:mirror-main-distro-default-smithi/

It looks like sometimes a mirrored image's status is not captured in `rbd mirror pool status` when there are several images being mirrored. This can explain why without the above change, the workload `compare_mirror_image_alternate_primary.sh`, which involved only one mirrored image, passed.

Reply:
Have you confirmed that it's actually `rbd mirror pool status` that returns no output?  I think it might be a reentrancy issue with the helper: it seems like all `test_mirror_pool_status_verbose()` invocations redirect the output to the same file, which is truncated in the process.  It's possible that the truncate races with one of the later checks.

Reply:
You're right. That's the issue.

How about using a UUID in the names of files that store `mirror image status` or `mirror pool status`?
```
diff --git a/qa/workunits/rbd/rbd_mirror_helpers.sh b/qa/workunits/rbd/rbd_mirror_helpers.sh
index f7a3e5de0d9..ad7427fe97e 100755
--- a/qa/workunits/rbd/rbd_mirror_helpers.sh
+++ b/qa/workunits/rbd/rbd_mirror_helpers.sh
@@ -823,7 +823,7 @@ test_status_in_pool_dir()
     local description_pattern="$5"
     local service_pattern="$6"
 
-    local status_log=${TEMPDIR}/$(mkfname ${cluster}-${pool}-${image}.mirror_status)
+    local status_log=${TEMPDIR}/$(uuidgen).mirror_image_status
     CEPH_ARGS='' rbd --cluster ${cluster} mirror image status ${pool}/${image} |
         tee ${status_log} >&2
     grep "^  state: .*${state_pattern}" ${status_log} || return 1
@@ -856,7 +856,7 @@ test_mirror_pool_status_verbose()
     local state_pattern="$4"
     local prev_last_update="$5"
 
-    local status_log=${TEMPDIR}/$(mkfname ${cluster}-${pool}.mirror_status)
+    local status_log=${TEMPDIR}/$(uuidgen).mirror_pool_status
 
     rbd --cluster ${cluster} mirror pool status ${pool} --verbose --format xml \
         > ${status_log}
```

Reply:
> How about using a UUID in the names of files that store `mirror image status` or `mirror pool status`?

I think something like `mktemp XXXXXXXXXX.mirror_status` would be a bit more natural.

OTOH, I might be missing something, but I don't see a reason to use the file at all.  Why not just capture the output in a variable, potentially echoing it separately so that raw output makes it to the log?

Reply:
Using a variable makes sense. Thanks! Done.